### PR TITLE
feat(sdk/client): guess user namespace before using default namespace value

### PR DESCRIPTION
**Pull request by shawnzhu**
**Description of your changes:**
The current multi-user feature will force user to provide user namespace explicitly, which requires user to run either `client.set_user_namespace(<usernamespace>)` or using argument `client.create_experiment(name, namespace=<user_namespace>)` . this change will cover the use case that using SDK in a in-cluster notebook where it will guess user namespace by using below convention which was introduced to the notebook controller from https://github.com/kubeflow/kubeflow/pull/2620:

```SHELL
NB_PREFIX=/notebook/<user_namespace>/<notebook-instance-name>
```

So that it won't force all KFP samples to use `client.set_user_namespace()` when multi-user feature is released.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

### CODE REVIEW

**Comment by shawnzhu** [comment]
Supports https://github.com/kubeflow/manifests/issues/1364

File: sdk/python/kfp/_client.py [hunk 1]
```diff
@@ -255,8 +255,21 @@ def _load_context_setting_or_default(self):
         self._context_setting = json.load(f)
     else:
       self._context_setting = {
-        'namespace': '',
+        'namespace': self._guess_user_namespace_or_default(),
       }
+
+  def _guess_user_namespace_or_default(self):
+    """Guess user namespace from environment variable NB_PREFIX"""
```
Comment (Jeffwan) [inline sdk/python/kfp/_client.py:262]: This is good. Have same feeling, setting namespace explicitly is kind of unnecessary.

This implementation uses business logic and has the dependency on how to build jupyter notebook. 
Do you think it's better to use an application neutral way? 

1. check file `/var/run/secrets/kubernetes.io/serviceaccount/namespace` 
2. Change notebook side and already add ENN POD_NAMESPACE and here, we can just read `POD_NAMESPACE`

Comment (yanniszark) [reply sdk/python/kfp/_client.py:262]: I absolutely agree about using an application neutral way.
Using the notebook virtualservice path seems brittle.

Comment (Bobgy) [reply sdk/python/kfp/_client.py:262]: +1 to this suggestion

Comment (shawnzhu) [reply sdk/python/kfp/_client.py]: Good idea! addressed in https://github.com/kubeflow/pipelines/pull/4293/commits/8575a57fce9c4527e217f0eca32357141d3e5f77

**Comment by animeshsingh** [comment]
thanks @shawnzhu 
/lgtm

File: sdk/python/kfp/_client.py [hunk 2]
```diff
@@ -255,8 +255,19 @@ def _load_context_setting_or_default(self):
...
-        'namespace': '',
+        'namespace': self._guess_user_namespace_or_default(),
       }
+
+  def _guess_user_namespace_or_default(self):
+    """Guess user namespace from pod service account metadata"""
+    SA_NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+    default_namespace = ''
+
+    if os.path.isfile(SA_NAMESPACE_PATH):
```
Comment (Bobgy) [inline sdk/python/kfp/_client.py:266]: Sorry, I am not very familiar with python. Will this ever throw an exception? For example, no permission

Comment (shawnzhu) [reply sdk/python/kfp/_client.py:266]: No, it won't throw exception based on current implementation: https://github.com/python/cpython/blob/master/Lib/genericpath.py#L27-L33

E.g.:

```Python
$ ls /root/
ls: cannot open directory /root/: Permission denied
$ python
Python 2.7.5 (default, Aug  7 2019, 00:51:29) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> st = os.stat('/root/anyfile')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 13] Permission denied: '/root/file'
>>> os.path.isfile('/root/anyfile')
False
```

**Comment by Bobgy** [comment]
Looks like the change broke backward compatibility:
> HTTP response body: {"error":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references.","message":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references.","code":3,"details":[{"@type":"type.googleapis.com/api.Error","error_message":"In single-user mode, CreateExperimentRequest shouldn't contain resource references.","error_details":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references."}]}

These are copied from test failure.

**Comment by Bobgy** [comment]
I think KFP API server needs to provide an endpoint that exposes some cluster specific information:
e.g. what is expected mode, single/multi? What is ui_host? etc.

With these info, kfp sdk can auto configure more stuff for users.

**Comment by Ark-kun** [comment]
> I think KFP API server needs to provide an endpoint that exposes some cluster specific information:
> e.g. what is expected mode, single/multi? What is ui_host? etc.
> 
> With these info, kfp sdk can auto configure more stuff for users.

I agree. Also methods to get/set artifact location.

**Comment by Ark-kun** [comment]
```
integration-test-mc8z7-558845744: kfp_server_api.exceptions.ApiException: (400)
integration-test-mc8z7-558845744: Reason: Bad Request
integration-test-mc8z7-558845744: HTTP response headers: HTTPHeaderDict({'Content-Type': 'application/json', 'Trailer': 'Grpc-Trailer-Content-Type', 'Date': 'Tue, 04 Aug 2020 02:15:27 GMT', 'Transfer-Encoding': 'chunked'})
integration-test-mc8z7-558845744: HTTP response body: {"error":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references.","message":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references.","code":3,"details":[{"@type":"type.googleapis.com/api.Error","error_message":"In single-user mode, CreateExperimentRequest shouldn't contain resource references.","error_details":"Validate experiment request failed.: Invalid input error: In single-user mode, CreateExperimentRequest shouldn't contain resource references."}]}
```

**Comment by NikeNano** [comment]
> I think KFP API server needs to provide an endpoint that exposes some cluster specific information:
> e.g. what is expected mode, single/multi? What is ui_host? etc.
> 
> With these info, kfp sdk can auto configure more stuff for users.

Do we have a issue for this?

**Comment by Bobgy** [comment]
@NikeNano feel free to create one

**Comment by Bobgy** [comment]
I think we can close this, considering #4638 is almost ready to merge

**Pull request closed without merging**